common/connection.go

```diff
@@ -238,15 +250,25 @@ func (c *Connection) createSession(info *target.Info) (*Session, error) {
 }
 
 func (c *Connection) handleIOError(err error) {
-	c.logger.Errorf("cdp", "communicating with browser: %v", err)
+	// It's either a normal closure initiated by the browser, or we stopped the
+	// connection. In either case, disregard the error, but close all sessions
+	// and emit a ConnectionClose event, so that the browser event listener can
+	// be stopped.
+	if closing := c.isClosing(); websocket.IsCloseError(
+		err, websocket.CloseNormalClosure, websocket.CloseGoingAway,
+	) || closing {
+		c.logger.Debugf("cdp", "received IO error: %v, connection is closing: %v", err, closing)
+		c.closeAllSessions()
+		c.emit(EventConnectionClose, nil)
```

Comment:
```suggestion
		if closing {
			c.emit(EventConnectionClose, nil)
		}
```

From my local testing the error that we see in the CI as well is because this event gets emitted while the `browser.close` event is still being "done" but after it was already send - so that the actual connection is closed. 

While the transport for that works on to return that `browser.Close` has been finished, the event gets emitted and then that leads to the cancelation of the connection browser context in https://github.com/grafana/xk6-browser/blob/430e82c73115ee34b13721d364dd95c664a2c2e6/common/browser.go#L183-L185
https://github.com/grafana/xk6-browser/blob/430e82c73115ee34b13721d364dd95c664a2c2e6/common/browser.go#L198-L201

Reply:
Unfortunately as always I am getting other failures - some of which I have a patch for ... the others are timeouts which ... seem bad, but I am not certain if I have somehow fixed them in the last few minutes or they just do not happen at the moment :shrug:

Reply:
Also you still need to emit that event :facepalm: but after the close has happened. I did return to calling the connection.close method and am currently testing but ... likely we can just emit the event after the browser.Close :thinking: 

On the other hand maybe it should be emitted after we *actually* close the connectiong :shrug:. (given that is the name of the event ;))

Reply:
https://github.com/grafana/xk6-browser/pull/541/commits/f24516934a417c5ff1511e4308cbfa6c56c5c38b is what I mean

Reply:
:+1: Thanks for looking into this, and the fix.

Indeed, emitting `EventConnectionClose` could race and cancel the browser context too early.

I think we can't call `Connection.Close()` right after sending `Browser.close`, since we need to wait for incoming events like `Target.detachedFromTarget` to arrive and be processed. They're not critical in this case since we're shutting down and ending the iteration anyway, but we should handle those cleanly.

Otherwise, you can see this error with `XK6_BROWSER_LOG=trace`:

```
received IO error: read tcp 127.0.0.1:43488->127.0.0.1:40137: use of closed network connection, connection is closing: true
```

There's also some duplication with what's done in `handleIOError()`. E.g. we might emit `EventConnectionClose` twice.

But let's merge your PR and I'll try to clean it up.

Reply:
I opted to just move the `Connection.Close()` call after the browser process exits, which is unintuitive, but hopefully the comment in 5e8139e explains the reason. We could always move what's needed into a separate function and call it instead, but a) I couldn't decide what to name that function, and b) it might add some duplication, or change of logic, since we use `sync.Once` there.

I think this is fine, but let me know if you have a better idea.